src/Pool.sol

```diff
@@ -475,22 +397,55 @@ abstract contract Pool {
     /// Must be sorted by the receivers' addresses, deduplicated and without 0 amtPerSecs.
     /// @return withdrawn The withdrawn amount which should be sent to the user.
     /// Equal to `withdrawAmt` unless `WITHDRAW_ALL` is used.
-    function _updateAnySender(
-        Sender memory sender,
+    function _updateSender(
         SenderId memory senderId,
         uint128 topUpAmt,
         uint128 withdrawAmt,
         Receiver[] calldata currReceivers,
         Receiver[] calldata newReceivers
     ) internal returns (uint128 withdrawn) {
-        _assertCurrReceiversHash(sender.receiversHash, currReceivers);
+        Sender memory sender = _loadSender(senderId);
+        _assertCurrReceivers(sender, currReceivers);
+
         uint128 newAmtPerSec = _setReceiversHash(sender, newReceivers);
         uint128 currAmtPerSec = _totalAmtPerSec(currReceivers);
         uint64 currEndTime = _sendingEndTime(sender, currAmtPerSec);
         withdrawn = _updateSenderBalance(sender, topUpAmt, withdrawAmt, currAmtPerSec, currEndTime);
         sender.startTime = _currTimestamp();
         uint64 newEndTime = _sendingEndTime(sender, newAmtPerSec);
         _updateStreams(senderId, currReceivers, newReceivers, currEndTime, newEndTime);
+
+        _storeSender(senderId, sender);
+        _emitSenderUpdated(senderId, sender.startBalance, newReceivers);
+        _transfer(senderId.senderAddr, int128(withdrawn) - int128(topUpAmt));
+    }
+
+    function _loadSender(SenderId memory senderId) internal view returns (Sender memory) {
+        if (senderId.isSubSender) {
+            return subSenders[senderId.senderAddr][senderId.subSenderId];
```

Comment:
Did we compare the gas difference if we would store both in one global mapping based on the hash of `senderAddr` and `subSenderId`?

One pattern for upgradeability is to implement some kind of storage contract with getter and setter methods. 

Maybe we can discuss this a bit, seems like we heading a bit anyway towards this direction with `load` and `store` methods.

On the other hand it will increase the gas costs again, so i guess it is fine.

Reply:
The gas usage would increase a little. In the current setup a `sender` is stored at slot
```solidity
keccak256(
  concat(
    uint256(senderId.senderAddr),
    SENDER_MAPPING_SLOT
  )
)
```
while a sub-sender at slot
```solidity
keccak256(
  concat(
    senderId.subSenderId,
    keccak256(
      concat(
        uint256(senderId.senderAddr),
        SUB_SENDER_MAPPING_SLOT
      )
   )
)
```
IIUC what you're proposing would store a sender at:
```solidity
keccak256(
  concat(
    keccak256(
      concat(
        senderId.senderAddr,
        senderId.isSubSender,
        senderId.isSubSender ? sender.subSenderId : 0
      ),
      SUB_SENDER_MAPPING_SLOT
   )
)
```
For sub-senders it's pretty much the same thing, but for senders it hashes one time more.

To put it in context, hashing costs 30 gas + 6 gas per hashed word and we're accessing storage no more than twice per call. Obfuscating data structure is not worth any savings in this area.